### ISSUE TITLE
Tensor indexing fail on MPS backend. 

### ISSUE THREAD

**Issue by gad-imageinsight** [pytorch/pytorch#84229]
### 🐛 Describe the bug

Issue arises in running an inference in a YOLOV3 model when coordinates in the results need to be rescaled. Something funky happens with the tensor access on a -= operator. 

You'll need to set PYTORCH_ENABLE_MPS_FALLBACK=1 in the environment because of missing mps methods

```
import torch
model = torch.hub.load('ultralytics/yolov3', 'yolov3')
device = torch.device('mps')
model.to(device)
img = 'https://ultralytics.com/images/zidane.jpg'
model(img)
```

I put debugging code into .cache/torch/hub/ultralytics_yolov3_master/utils/general.py in the scale_coords method.
The issue arises in the lines
```
coords[:,[0,2]] -= pad[0] 
coords[:,[1,3]] -= pad[1] 
```

For this image the rescaling is going from [384, 640] -> (720, 1280) giving a gain of 0.5 and a padding of (0.0,12.0) in 
that function. Before these two lines, the coords tensor is:
```
tensor([[ 71.47987, 112.43367, 531.28778, 369.58679],
        [371.92389,  33.11592, 563.16974, 368.19769],
        [216.71069, 228.54417, 260.64929, 370.04932],
        [495.18076, 170.43976, 511.94131, 220.99673]], device='mps:0')
```
After the x padding
```
tensor([[ 71.47987, 112.43367, 531.28778, 369.58679],
        [563.16974, 368.19769, 216.71069, 228.54417],
        [495.18076, 170.43976, 511.94131, 220.99673],
        [495.18076, 170.43976, 511.94131, 220.99673]], device='mps:0')
```
The first two entries in the second row have been replaced by the last two, and those by the first 2 in the third row. 
The third row is now a duplicate of the fourth.

After the y padding
```
tensor([[ 71.47987, 100.43367, 531.28778, 357.58679],
        [216.71069, 216.54417, 495.18076, 158.43976],
        [495.18076, 158.43976, 511.94131, 208.99673],
        [495.18076, 170.43976, 511.94131, 220.99673]], device='mps:0')
```
The same thing has happened again, except the last line hasn't gotten the padding done to it. 

For comparison, this is the same calculation, but on the GPU after the y padding:
```
tensor([[ 71.47993, 100.43364, 531.28772, 357.58685],
        [371.92389,  21.11584, 563.16980, 356.19778],
        [216.71071, 216.54416, 260.64926, 358.04935],
        [495.18079, 158.43976, 511.94128, 208.99673]])
```
which is what is expected.

So I did some sleuthing and stripped it down, and this reproduces the error:
```
import numpy as np
from torch import tensor
import torch

import sys

# you'll need to change this to whatever is appropriate on your system
sys.path.append('/Users/gad/.cache/torch/hub/ultralytics_yolov3_master')
from utils.general import scale_coords

print('numpy', np.__version__)
print('pytorch', torch.__version__)

device = torch.device("mps")

coords = np.array([[7.14799e+01, 1.12434e+02, 5.31288e+02, 3.69587e+02, 9.36926e-01, 0.00000e+00],
        [3.71924e+02, 3.31158e+01, 5.63170e+02, 3.68198e+02, 9.16234e-01, 0.00000e+00],
        [2.16711e+02, 2.28544e+02, 2.60649e+02, 3.70049e+02, 8.61753e-01, 2.70000e+01],
        [4.95181e+02, 1.70440e+02, 5.11941e+02, 2.20997e+02, 3.32481e-01, 2.70000e+01]],dtype=np.float32)

img1_shape = [384,640]
img0_shape = (1080, 1920)

oncpu = tensor(coords)
outc = scale_coords(img1_shape, oncpu[:,:4], img0_shape)
print(f'on cpu {outc}')

ongpu = tensor(coords, device=device )
outg = scale_coords(img1_shape, ongpu[:,:4] , img0_shape) 
print(f'on gpu {outg}')
```

with output:
```
on cpu tensor([[ 214.43968,  301.30197, 1593.86401, 1072.76099],
        [1115.77197,   63.34740, 1689.50989, 1068.59399],
        [ 650.13300,  649.63202,  781.94696, 1074.14697],
        [1485.54297,  475.32001, 1535.82300,  626.99097]])
on gpu tensor([[ 214.43970,  301.30200, 1593.86401, 1072.76099],
        [ 650.13300,  649.63202, 1485.54297,  475.32001],
        [1485.54297,  475.32001, 1535.82300,  626.99097],
        [1485.54297,  511.32001, 1535.82300,  662.99097]], device='mps:0')
```

### Versions

PyTorch version: 1.13.0.dev20220829
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==1.13.0.dev20220829
[pip3] torchaudio==0.13.0.dev20220829
[pip3] torchvision==0.14.0.dev20220829
[conda] numpy                     1.22.4           py39h7df2422_0    conda-forge
[conda] pytorch                   1.13.0.dev20220829         py3.9_0    pytorch-nightly
[conda] torchaudio                0.13.0.dev20220829        py39_cpu    pytorch-nightly
[conda] torchvision               0.14.0.dev20220829        py39_cpu    pytorch-nightly

cc @kulinseth @albanD

**Comment by leovinus2001** [comment]
I do not want to distract from the original test case, but as the title mentions "indexing fails on MPS backend", may I point out that some simple MPS indexing like the code below fails on my machine (Intel iMac, Torch 1.13.0a0+gitf4f54c7 from 2 days ago). That seemed to me a possibly related issue.

When you just print [0][0] is works fine but [100][100] causes a segfault of Python after 0 or more iterations. It prints the message

>torch/_tensor_str.py:114: UserWarning: The operator 'aten::masked_select' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at aten/src/ATen/mps/MPSFallback.mm:11.) nonzero_finite_vals = torch.masked_select(

and crashes at

>4   libtorch_cpu.dylib            	       0x1180243cc at::mps::MPSStream::copy_and_sync(id<MTLBuffer>, objc_object<MTLBuffer>, unsigned long, unsigned long, unsigned long, bool) + 108

per attached crash.log
```
import torch
    
print (torch.__version__)

for ii in range(10000):
    fa = torch.ones(499, 526, device='mps')

    #print (fa[0][0]) # Fine
    print (fa[100][100]) # Segfault after a random number of iterations
```
[crash.log](https://github.com/pytorch/pytorch/files/9463704/crash.log)

**Comment by DenisVieriu97** [comment]
Hi @gad-imageinsight - I've tried reproducing this with latest nightly, but with no success.
Could you please try the latest pytorch nightly build (1.13.0.dev20220930) and let me know if you're still seeing the issue? Thanks!

**Comment by gad-imageinsight** [comment]
I tried the test program above with build 1.13.0.dev20221002. Results:
```
on cpu tensor([[ 214.43968,  301.30197, 1593.86401, 1072.76099],
        [1115.77197,   63.34740, 1689.50989, 1068.59399],
        [ 650.13300,  649.63202,  781.94696, 1074.14697],
        [1485.54297,  475.32001, 1535.82300,  626.99097]])
on gpu tensor([[ 214.43970,  301.30200, 1593.86401, 1072.76099],
        [1115.77197,   63.34740, 1689.51001, 1068.59399],
        [ 650.13300,  649.63202,  781.94696, 1074.14697],
        [1485.54297,  475.32001, 1535.82300,  626.99097]], device='mps:0')
```
Confirming that it has been fixed! 
Thanks very much.
